### PR TITLE
[SPARK-23033][SS] Don't use task level retry for continuous processing

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Continuous processing tasks will fail on any attempt number greater than 0. ContinuousExecution will catch these failures and restart globally from the last recorded checkpoints.
## How was this patch tested?
unit test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 1]
```diff
@@ -219,6 +199,37 @@ class ContinuousSuite extends ContinuousSuiteBase {
...
+    spark.sparkContext.addSparkListener(listener)
+
+    testStream(df, useV2Sink = true)(
+      StartStream(Trigger.Continuous(100)),
+      Execute(waitForRateSourceTriggers(_, 2)),
+      Execute { _ =>
+        eventually(timeout(streamingTimeout)) { assert(taskId != -1) }
+        spark.sparkContext.killTaskAttempt(taskId)
+      },
+      Execute(waitForRateSourceTriggers(_, 4)),
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: can you explain the logic behind this test? what does this test do?

Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: It kills an arbitrary task, and checks that query execution continues onward unaffected.

I've added a check that the run ID has changed, confirming that the retry was indeed made at the ContinuousExecution level.

**Comment by jose-torres** [comment]
The hang in test build #85939 was a test issue in ContinuousStressSuite, which I could reproduce locally by bumping up the rows per second.

When the rate of incoming data is too high, the query execution *does* make progress -  but the answer checking operation in the test is expensive. And the rate source was continuing to run as it happened. Since the executors are local, an overload means the answer check will take unreasonably long to finish.

I've fixed this by stopping the rate source before checking the answer in the stress tests.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 2]
```diff
@@ -219,6 +201,44 @@ class ContinuousSuite extends ContinuousSuiteBase {
       StopStream)
   }
 
+  test("kill task") {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: this test does not verify killing tasks :) it verifies "task failure stops the query" or "task failure restarts the query"

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 3]
```diff
@@ -219,6 +201,42 @@ class ContinuousSuite extends ContinuousSuiteBase {
...
+        eventually(timeout(streamingTimeout)) {
+          assert(query.exception.isDefined)
+        }
+        assert(
+          query.exception.get.getCause != null &&
+          query.exception.get.getCause.getCause != null &&
+          query.exception.get.getCause.getCause.getCause.isInstanceOf[ContinuousTaskRetryException])
+      })
+
+    spark.sparkContext.removeSparkListener(listener)
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: put this in a finally clause.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 4]
```diff
@@ -258,13 +276,9 @@ class ContinuousStressSuite extends ContinuousSuiteBase {
...
       Execute(waitForRateSourceTriggers(_, 201)),
       IncrementEpoch(),
-      Execute { query =>
-        val data = query.sink.asInstanceOf[MemorySinkV2].allData
-        val vals = data.map(_.getLong(0)).toSet
-        assert(scala.Range(0, 25000).forall { i =>
-          vals.contains(i)
-        })
-      })
+      StopStream,
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: What is this for?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 5]
```diff
@@ -280,6 +294,7 @@ class ContinuousStressSuite extends ContinuousSuiteBase {
       AwaitEpoch(0),
       Execute(waitForRateSourceTriggers(_, 201)),
       IncrementEpoch(),
+      StopStream,
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: Why are these needed?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala [hunk 6]
```diff
@@ -219,6 +201,42 @@ class ContinuousSuite extends ContinuousSuiteBase {
...
+    spark.sparkContext.addSparkListener(listener)
+
+    testStream(df, useV2Sink = true)(
+      StartStream(Trigger.Continuous(100)),
+      Execute(waitForRateSourceTriggers(_, 2)),
+      Execute { query =>
+        // Wait until a task is started, then kill its first attempt.
+        eventually(timeout(streamingTimeout)) { assert(taskId != -1) }
+        spark.sparkContext.killTaskAttempt(taskId)
+        eventually(timeout(streamingTimeout)) {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: Can this be checked with a "ExpectFailure" test? Better to test using the same harness that is used for microbatch so that we are sure they failure behavior is the same.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTaskRetryException.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.continuous
+
+import org.apache.spark.SparkException
+
+class ContinuousTaskRetryException
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTaskRetryException.scala]: Add docs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousDataSourceRDDIter.scala [hunk 8]
```diff
@@ -52,6 +52,10 @@ class ContinuousDataSourceRDD(
   }
 
   override def compute(split: Partition, context: TaskContext): Iterator[UnsafeRow] = {
+    if (context.attemptNumber() != 0) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousDataSourceRDDIter.scala]: Add comments on what this is.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala (hunk 4 above)
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala]: This is the earlier comment about the overloaded failure mode this PR exposed.

**Comment by jose-torres** [comment]
The problem in the test runs above was also a test issue. The new code changed the synchronization such that the row wasn't written when the test expected; I verified manually that the failing test doesn't actually enter the attemptNumber != 0 branch.

**Comment by jose-torres** [comment]
The most recent test build failure is from an earlier commit which I think is obsoleted. I think #86007 is correct but we should retest this please to confirm.

**Comment by jose-torres** [comment]
retest this please

**Comment by jose-torres** [comment]
retest this please

**Comment by tdas** [comment]
Fix merge conflicts. And add [SS} to the title of this PR.

**Comment by tdas** [comment]
LGTM. Merging this to master and 2.3
